### ISSUE TITLE
Paragraphs are not ending properly

### ISSUE THREAD

**Issue by pard68** [darakian/mini_markdown#19]
Paragraphs are not being rendered in HTML correctly because paragraph tokens only seem to be ending at the very end of the parsing and not when a paragraph ends.

Minimum code to reproduce problem:
```rust
use mini_markdown;

fn main() {
    let md: &str = "Paragraph 1.\n\n```\nBlock text should end a paragraph.\n```\n\nThis is paragraph two.\n\n## Heading\n\nParagraph the third.";
    println!("{:#?}", mini_markdown::lex(md));
}
```

Actual output:
```html
<p>Paragraph 1.
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Block text should end a paragraph.
</code></pre></div></div>
This is paragraph two.
<h2 id="heading">Heading</h2>

Paragraph the third.</p>
```

Expected output:
```html
<p>Paragraph 1.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Block text should end a paragraph.
</code></pre></div></div>
<p>This is paragraph two.</p>
<h2 id="heading">Heading</h2>

<p>Paragraph the third.</p>
```

Looks like there is also some issues, potentially, with spacing though that shouldn't matter as much as getting the `<p>` tag to close correctly.

It would appear that the lexer has a token for a `ParagraphBreak`, however it's never implemented by the `lex` function. The above markdown example puts out the following lex:

```
[
    Plaintext(
        "Paragraph 1.",
    ),
    Newline,
    CodeBlock(
        "Block text should end a paragraph.\n",
        "plaintext",
    ),
    Newline,
    Plaintext(
        "This is paragraph two.",
    ),
    Newline,
    Header(
        2,
        "Heading",
        None,
    ),
    Newline,
    Plaintext(
        "Paragraph the third.",
    ),
]
```

I would expect to see each `Plaintext` item followed by a `ParagraphBreak`. Not sure if this missing token is actually the problem though, as it also seems that the `parse` function is trying to handle closing `p` tags without the use of the `ParagraphBreak` token.

**Comment by darakian** [comment]
Hey there 👋 

Thanks for the issues. Paragraph handling is something that certainly needs work and ya, the `ParagraphBreak` token was an idea that didn't really work out since the plaintext blocks themselves came to define a beginning and an end to the paragraphs. You're right that this issue and #20 are coupled together. I've made a branch which should address your issues (it at least addresses your test cases). Would you mind giving it a spin? 

https://github.com/darakian/mini_markdown/pull/21

**Comment by pard68** [comment]
Sorry for the delay. My tests do work on that branch, haven't tested with much beyond the examples I provided above though. Thanks!

**Comment by darakian** [comment]
Happy to do it and thanks for the feedback :)

I'm going to close out both issues and I'll make a new release by next weekend.

**Comment by darakian** [comment]
Resolved in #21
